{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# !pip install flask\n",
    "# !pip install line-bot-sdk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n\\n整體功能描述\\n\\n應用伺服器主結構樣板\\n    用來定義使用者傳送消息時，應該傳到哪些方法上\\n        比如收到文字消息時，轉送到文字專屬處理方法\\n\\n消息專屬方法定義\\n    當收到文字消息，從資料夾內提取消息，並進行回傳。\\n    \\n啟動應用伺服器    \\n\\n'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "\n",
    "整體功能描述\n",
    "\n",
    "應用伺服器主結構樣板\n",
    "    用來定義使用者傳送消息時，應該傳到哪些方法上\n",
    "        比如收到文字消息時，轉送到文字專屬處理方法\n",
    "\n",
    "消息專屬方法定義\n",
    "    當收到文字消息，從資料夾內提取消息，並進行回傳。\n",
    "    \n",
    "啟動應用伺服器    \n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "Application 主架構\n",
    "\n",
    "'''\n",
    "\n",
    "# 引用Web Server套件\n",
    "from flask import Flask, request, abort\n",
    "\n",
    "# 從linebot 套件包裡引用 LineBotApi 與 WebhookHandler 類別\n",
    "from linebot import (\n",
    "    LineBotApi, WebhookHandler\n",
    ")\n",
    "\n",
    "# 引用無效簽章錯誤\n",
    "from linebot.exceptions import (\n",
    "    InvalidSignatureError\n",
    ")\n",
    "\n",
    "# 載入json處理套件\n",
    "import json\n",
    "\n",
    "# 載入基礎設定檔\n",
    "secretFileContentJson=json.load(open(\"./line_secret_key\",\"r\",encoding=\"utf-8\"))\n",
    "\n",
    "# 設定Server啟用細節\n",
    "app = Flask(__name__,static_url_path = \"/material\" , static_folder = \"./material/\")\n",
    "\n",
    "# 生成實體物件\n",
    "line_bot_api = LineBotApi(secretFileContentJson.get(\"channel_access_token\"))\n",
    "handler = WebhookHandler(secretFileContentJson.get(\"secret_key\"))\n",
    "\n",
    "# 啟動server對外接口，使Line能丟消息進來\n",
    "@app.route(\"/\", methods=['POST'])\n",
    "def callback():\n",
    "    # get X-Line-Signature header value\n",
    "    signature = request.headers['X-Line-Signature']\n",
    "\n",
    "    # get request body as text\n",
    "    body = request.get_data(as_text=True)\n",
    "    app.logger.info(\"Request body: \" + body)\n",
    "\n",
    "    # handle webhook body\n",
    "    try:\n",
    "        handler.handle(body, signature)\n",
    "    except InvalidSignatureError:\n",
    "        abort(400)\n",
    "\n",
    "    return 'OK'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "消息判斷器\n",
    "\n",
    "讀取指定的json檔案後，把json解析成不同格式的SendMessage\n",
    "\n",
    "讀取檔案，\n",
    "把內容轉換成json\n",
    "將json轉換成消息\n",
    "放回array中，並把array傳出。\n",
    "\n",
    "'''\n",
    "\n",
    "# 引用會用到的套件\n",
    "from linebot.models import (\n",
    "    ImagemapSendMessage,TextSendMessage,ImageSendMessage,LocationSendMessage,FlexSendMessage,VideoSendMessage\n",
    ")\n",
    "\n",
    "from linebot.models.template import (\n",
    "    ButtonsTemplate,CarouselTemplate,ConfirmTemplate,ImageCarouselTemplate\n",
    "    \n",
    ")\n",
    "\n",
    "from linebot.models.template import *\n",
    "\n",
    "def detect_json_array_to_new_message_array(fileName):\n",
    "    \n",
    "    #開啟檔案，轉成json\n",
    "    with open(fileName,'r',encoding='utf8') as f:\n",
    "        jsonArray = json.load(f)\n",
    "    \n",
    "    # 解析json\n",
    "    returnArray = []\n",
    "    for jsonObject in jsonArray:\n",
    "\n",
    "        # 讀取其用來判斷的元件\n",
    "        message_type = jsonObject.get('type')\n",
    "        \n",
    "        # 轉換\n",
    "        if message_type == 'text':\n",
    "            returnArray.append(TextSendMessage.new_from_json_dict(jsonObject))\n",
    "        elif message_type == 'imagemap':\n",
    "            returnArray.append(ImagemapSendMessage.new_from_json_dict(jsonObject))\n",
    "        elif message_type == 'template':\n",
    "            returnArray.append(TemplateSendMessage.new_from_json_dict(jsonObject))\n",
    "        elif message_type == 'image':\n",
    "            returnArray.append(ImageSendMessage.new_from_json_dict(jsonObject))\n",
    "        elif message_type == 'sticker':\n",
    "            returnArray.append(StickerSendMessage.new_from_json_dict(jsonObject))  \n",
    "        elif message_type == 'audio':\n",
    "            returnArray.append(AudioSendMessage.new_from_json_dict(jsonObject))  \n",
    "        elif message_type == 'location':\n",
    "            returnArray.append(LocationSendMessage.new_from_json_dict(jsonObject))\n",
    "        elif message_type == 'flex':\n",
    "            returnArray.append(FlexSendMessage.new_from_json_dict(jsonObject))  \n",
    "        elif message_type == 'video':\n",
    "            returnArray.append(VideoSendMessage.new_from_json_dict(jsonObject))    \n",
    "\n",
    "\n",
    "    # 回傳\n",
    "    return returnArray"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "handler處理關注消息\n",
    "\n",
    "用戶關注時，讀取 素材 -> 關注 -> reply.json\n",
    "\n",
    "將其轉換成可寄發的消息，傳回給Line\n",
    "\n",
    "'''\n",
    "\n",
    "# 引用套件\n",
    "from linebot.models import (\n",
    "    FollowEvent\n",
    ")\n",
    "\n",
    "# 關注事件處理\n",
    "@handler.add(FollowEvent)\n",
    "def process_follow_event(event):\n",
    "    \n",
    "    # 讀取並轉換\n",
    "    result_message_array =[]\n",
    "    replyJsonPath = \"material/follow/reply.json\"\n",
    "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
    "\n",
    "    # 消息發送\n",
    "    line_bot_api.reply_message(\n",
    "        event.reply_token,\n",
    "        result_message_array\n",
    "    )\n",
    "\n",
    "    # 從follow資料夾內，取出圖文選單id，並告知line，綁定用戶\n",
    "    linkRichMenuId = open(\"material/follow/rich_menu_id\", 'r').read()\n",
    "    line_bot_api.link_rich_menu_to_user(event.source.user_id, linkRichMenuId)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "handler處理文字消息\n",
    "\n",
    "收到用戶回應的文字消息，\n",
    "按文字消息內容，往素材資料夾中，找尋以該內容命名的資料夾，讀取裡面的reply.json\n",
    "\n",
    "轉譯json後，將消息回傳給用戶\n",
    "\n",
    "'''\n",
    "\n",
    "# 引用套件\n",
    "from linebot.models import (\n",
    "    MessageEvent, TextMessage\n",
    ")\n",
    "\n",
    "# 文字消息處理\n",
    "@handler.add(MessageEvent,message=TextMessage)\n",
    "def process_text_message(event):\n",
    "\n",
    "    # 讀取本地檔案，並轉譯成消息\n",
    "    result_message_array =[]\n",
    "    replyJsonPath = \"material/\"+event.message.text+\"/reply.json\"\n",
    "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
    "\n",
    "    # 發送\n",
    "    line_bot_api.reply_message(\n",
    "        event.reply_token,\n",
    "        result_message_array\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "handler處理Postback Event\n",
    "\n",
    "載入功能選單與啟動特殊功能\n",
    "\n",
    "解析postback的data，並按照data欄位判斷處理\n",
    "\n",
    "現有三個欄位\n",
    "menu, folder, tag\n",
    "\n",
    "若folder欄位有值，則\n",
    "    讀取其reply.json，轉譯成消息，並發送\n",
    "\n",
    "若menu欄位有值，則\n",
    "    讀取其rich_menu_id，並取得用戶id，將用戶與選單綁定\n",
    "    讀取其reply.json，轉譯成消息，並發送\n",
    "\n",
    "'''\n",
    "from linebot.models import (\n",
    "    PostbackEvent\n",
    ")\n",
    "\n",
    "from urllib.parse import parse_qs \n",
    "\n",
    "@handler.add(PostbackEvent)\n",
    "def process_postback_event(event):\n",
    "    \n",
    "    query_string_dict = parse_qs(event.postback.data)\n",
    "    \n",
    "    print(query_string_dict)\n",
    "    if 'folder' in query_string_dict:\n",
    "    \n",
    "        result_message_array =[]\n",
    "\n",
    "        replyJsonPath = 'material/'+query_string_dict.get('folder')[0]+\"/reply.json\"\n",
    "        result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
    "  \n",
    "        line_bot_api.reply_message(\n",
    "            event.reply_token,\n",
    "            result_message_array\n",
    "        )\n",
    "    elif 'menu' in query_string_dict:\n",
    " \n",
    "        linkRichMenuId = open(\"material/\"+query_string_dict.get('menu')[0]+'/rich_menu_id', 'r').read()\n",
    "        line_bot_api.link_rich_menu_to_user(event.source.user_id,linkRichMenuId)\n",
    "        \n",
    "        replyJsonPath = 'material/'+query_string_dict.get('menu')[0]+\"/reply.json\"\n",
    "        result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
    "  \n",
    "        line_bot_api.reply_message(\n",
    "            event.reply_token,\n",
    "            result_message_array\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "若收到圖片消息時，\n",
    "\n",
    "先回覆用戶文字消息，並從Line上將照片拿回。\n",
    "\n",
    "'''\n",
    "from linebot.models import (\n",
    "    MessageEvent, TextMessage, TextSendMessage,ImageMessage\n",
    ")\n",
    "\n",
    "\n",
    "# import tmark_input_word as tiw\n",
    "\n",
    "# saveAns = ''\n",
    "\n",
    "@handler.add(MessageEvent, message=ImageMessage)\n",
    "def handle_message(event):\n",
    "#     global saveAns\n",
    "    \n",
    "    line_bot_api.reply_message(\n",
    "        event.reply_token, \n",
    "        TextSendMessage(text='照片已上傳成功~' + '\\n 等待模型回傳結果~\\n並輸入(看結果)'))\n",
    "    line_bot_api.reply_message(\n",
    "        event.reply_token,\n",
    "        TextSendMessage(text='看結果'))   \n",
    "    message_content = line_bot_api.get_message_content(event.message.id)\n",
    "\n",
    "    \n",
    "    with open('../images/'+event.message.id+'.jpg', 'wb') as fd:\n",
    "        for chunk in message_content.iter_content():\n",
    "            fd.write(chunk)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * Serving Flask app \"__main__\" (lazy loading)\n",
      " * Environment: production\n",
      "   WARNING: This is a development server. Do not use it in a production deployment.\n",
      "   Use a production WSGI server instead.\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " * Running on http://0.0.0.0:5000/ (Press CTRL+C to quit)\n",
      "127.0.0.1 - - [27/Jan/2021 22:33:35] \"\u001b[37mPOST / HTTP/1.1\u001b[0m\" 200 -\n",
      "127.0.0.1 - - [27/Jan/2021 22:33:45] \"\u001b[37mPOST / HTTP/1.1\u001b[0m\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'menu': ['rich_menu_fn']}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-01-27 22:33:53,121] ERROR in app: Exception on / [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-3-55911e9defe0>\", line 45, in callback\n",
      "    handler.handle(body, signature)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 260, in handle\n",
      "    self.__invoke_func(func, event, payload)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 272, in __invoke_func\n",
      "    func(event)\n",
      "  File \"<ipython-input-7-361c6385474e>\", line 49, in process_postback_event\n",
      "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
      "  File \"<ipython-input-4-5a5d33e6753e>\", line 29, in detect_json_array_to_new_message_array\n",
      "    with open(fileName,'r',encoding='utf8') as f:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'material/rich_menu_fn/reply.json'\n",
      "127.0.0.1 - - [27/Jan/2021 22:33:53] \"\u001b[35m\u001b[1mPOST / HTTP/1.1\u001b[0m\" 500 -\n",
      "[2021-01-27 22:33:54,156] ERROR in app: Exception on / [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-3-55911e9defe0>\", line 45, in callback\n",
      "    handler.handle(body, signature)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 260, in handle\n",
      "    self.__invoke_func(func, event, payload)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 272, in __invoke_func\n",
      "    func(event)\n",
      "  File \"<ipython-input-6-976e9f89b633>\", line 24, in process_text_message\n",
      "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
      "  File \"<ipython-input-4-5a5d33e6753e>\", line 29, in detect_json_array_to_new_message_array\n",
      "    with open(fileName,'r',encoding='utf8') as f:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'material/圖文/reply.json'\n",
      "127.0.0.1 - - [27/Jan/2021 22:33:54] \"\u001b[35m\u001b[1mPOST / HTTP/1.1\u001b[0m\" 500 -\n",
      "[2021-01-27 22:34:08,630] ERROR in app: Exception on / [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-3-55911e9defe0>\", line 45, in callback\n",
      "    handler.handle(body, signature)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 260, in handle\n",
      "    self.__invoke_func(func, event, payload)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 272, in __invoke_func\n",
      "    func(event)\n",
      "  File \"<ipython-input-8-f86eec905671>\", line 30, in handle_message\n",
      "    with open('../images/'+event.message.id+'.jpg', 'wb') as fd:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '../images/13455613216166.jpg'\n",
      "127.0.0.1 - - [27/Jan/2021 22:34:08] \"\u001b[35m\u001b[1mPOST / HTTP/1.1\u001b[0m\" 500 -\n",
      "127.0.0.1 - - [27/Jan/2021 22:34:14] \"\u001b[37mPOST / HTTP/1.1\u001b[0m\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'menu': ['rich_menu_fn']}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-01-27 22:34:23,546] ERROR in app: Exception on / [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-3-55911e9defe0>\", line 45, in callback\n",
      "    handler.handle(body, signature)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 260, in handle\n",
      "    self.__invoke_func(func, event, payload)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 272, in __invoke_func\n",
      "    func(event)\n",
      "  File \"<ipython-input-7-361c6385474e>\", line 49, in process_postback_event\n",
      "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
      "  File \"<ipython-input-4-5a5d33e6753e>\", line 29, in detect_json_array_to_new_message_array\n",
      "    with open(fileName,'r',encoding='utf8') as f:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'material/rich_menu_fn/reply.json'\n",
      "127.0.0.1 - - [27/Jan/2021 22:34:23] \"\u001b[35m\u001b[1mPOST / HTTP/1.1\u001b[0m\" 500 -\n",
      "[2021-01-27 22:34:24,576] ERROR in app: Exception on / [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-3-55911e9defe0>\", line 45, in callback\n",
      "    handler.handle(body, signature)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 260, in handle\n",
      "    self.__invoke_func(func, event, payload)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 272, in __invoke_func\n",
      "    func(event)\n",
      "  File \"<ipython-input-6-976e9f89b633>\", line 24, in process_text_message\n",
      "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
      "  File \"<ipython-input-4-5a5d33e6753e>\", line 29, in detect_json_array_to_new_message_array\n",
      "    with open(fileName,'r',encoding='utf8') as f:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'material/圖文/reply.json'\n",
      "127.0.0.1 - - [27/Jan/2021 22:34:24] \"\u001b[35m\u001b[1mPOST / HTTP/1.1\u001b[0m\" 500 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'menu': ['rich_menu_index']}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-01-27 22:34:26,839] ERROR in app: Exception on / [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-3-55911e9defe0>\", line 45, in callback\n",
      "    handler.handle(body, signature)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 260, in handle\n",
      "    self.__invoke_func(func, event, payload)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 272, in __invoke_func\n",
      "    func(event)\n",
      "  File \"<ipython-input-7-361c6385474e>\", line 49, in process_postback_event\n",
      "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
      "  File \"<ipython-input-4-5a5d33e6753e>\", line 29, in detect_json_array_to_new_message_array\n",
      "    with open(fileName,'r',encoding='utf8') as f:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'material/rich_menu_index/reply.json'\n",
      "127.0.0.1 - - [27/Jan/2021 22:34:26] \"\u001b[35m\u001b[1mPOST / HTTP/1.1\u001b[0m\" 500 -\n",
      "[2021-01-27 22:34:27,830] ERROR in app: Exception on / [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-3-55911e9defe0>\", line 45, in callback\n",
      "    handler.handle(body, signature)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 260, in handle\n",
      "    self.__invoke_func(func, event, payload)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 272, in __invoke_func\n",
      "    func(event)\n",
      "  File \"<ipython-input-6-976e9f89b633>\", line 24, in process_text_message\n",
      "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
      "  File \"<ipython-input-4-5a5d33e6753e>\", line 29, in detect_json_array_to_new_message_array\n",
      "    with open(fileName,'r',encoding='utf8') as f:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'material/回目錄/reply.json'\n",
      "127.0.0.1 - - [27/Jan/2021 22:34:27] \"\u001b[35m\u001b[1mPOST / HTTP/1.1\u001b[0m\" 500 -\n",
      "[2021-01-27 22:34:29,313] ERROR in app: Exception on / [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-3-55911e9defe0>\", line 45, in callback\n",
      "    handler.handle(body, signature)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 260, in handle\n",
      "    self.__invoke_func(func, event, payload)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 272, in __invoke_func\n",
      "    func(event)\n",
      "  File \"<ipython-input-7-361c6385474e>\", line 49, in process_postback_event\n",
      "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
      "  File \"<ipython-input-4-5a5d33e6753e>\", line 29, in detect_json_array_to_new_message_array\n",
      "    with open(fileName,'r',encoding='utf8') as f:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'material/rich_menu_ex/reply.json'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'menu': ['rich_menu_ex']}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Jan/2021 22:34:29] \"\u001b[35m\u001b[1mPOST / HTTP/1.1\u001b[0m\" 500 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'menu': ['rich_menu_index']}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2021-01-27 22:34:35,431] ERROR in app: Exception on / [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-3-55911e9defe0>\", line 45, in callback\n",
      "    handler.handle(body, signature)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 260, in handle\n",
      "    self.__invoke_func(func, event, payload)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 272, in __invoke_func\n",
      "    func(event)\n",
      "  File \"<ipython-input-7-361c6385474e>\", line 49, in process_postback_event\n",
      "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
      "  File \"<ipython-input-4-5a5d33e6753e>\", line 29, in detect_json_array_to_new_message_array\n",
      "    with open(fileName,'r',encoding='utf8') as f:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'material/rich_menu_index/reply.json'\n",
      "127.0.0.1 - - [27/Jan/2021 22:34:35] \"\u001b[35m\u001b[1mPOST / HTTP/1.1\u001b[0m\" 500 -\n",
      "[2021-01-27 22:34:36,225] ERROR in app: Exception on / [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 2447, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1952, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1821, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1950, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\flask\\app.py\", line 1936, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-3-55911e9defe0>\", line 45, in callback\n",
      "    handler.handle(body, signature)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 260, in handle\n",
      "    self.__invoke_func(func, event, payload)\n",
      "  File \"c:\\users\\uesr01\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\linebot\\webhook.py\", line 272, in __invoke_func\n",
      "    func(event)\n",
      "  File \"<ipython-input-6-976e9f89b633>\", line 24, in process_text_message\n",
      "    result_message_array = detect_json_array_to_new_message_array(replyJsonPath)\n",
      "  File \"<ipython-input-4-5a5d33e6753e>\", line 29, in detect_json_array_to_new_message_array\n",
      "    with open(fileName,'r',encoding='utf8') as f:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'material/上一頁/reply.json'\n",
      "127.0.0.1 - - [27/Jan/2021 22:34:36] \"\u001b[35m\u001b[1mPOST / HTTP/1.1\u001b[0m\" 500 -\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "\n",
    "Application 運行（開發版）\n",
    "\n",
    "'''\n",
    "if __name__ == \"__main__\":\n",
    "    app.run(host='0.0.0.0')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "Application 運行（heroku版）\n",
    "\n",
    "'''\n",
    "\n",
    "# import os\n",
    "# if __name__ == \"__main__\":\n",
    "#     app.run(host='0.0.0.0',port=os.environ['PORT'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
